keras_nlp/models/deberta/disentangled_self_attention.py

```diff
@@ -0,0 +1,376 @@
+
+        log_pos = _get_log_pos(abs_pos, mid)
+
+        bucket_pos = tf.where(
+            condition=abs_pos <= mid,
+            x=rel_pos,
+            y=log_pos * sign,
+        )
+        bucket_pos = tf.cast(bucket_pos, dtype=tf.int64)
+
+        return bucket_pos
+
+    def _get_rel_pos(self, num_positions):
+        ids = tf.range(num_positions, dtype=tf.int64)
+        query_ids = ids[:, tf.newaxis]
+        # query_ids = tf.repeat(query_ids, repeats=num_positions, axis=1)
+        key_ids = ids[tf.newaxis, :]
+        key_ids = tf.repeat(key_ids, repeats=num_positions, axis=0)
+
+        rel_pos = query_ids - key_ids
+        rel_pos = self._make_log_bucket_position(rel_pos)
+
+        rel_pos = rel_pos[tf.newaxis, tf.newaxis, :, :]
+        return rel_pos
+
+    def _compute_disentangled_attention(
+        self,
+        query,
+        key,
+        rel_embeddings,
+    ):
+        """Computes relative attention scores (p2c and c2p)."""
+
+        batch_size = tf.shape(query)[0]
+        num_positions = tf.shape(query)[1]
+
+        rel_pos = self._get_rel_pos(num_positions)
+
+        rel_attn_span = self.bucket_size
+        score = 0
+
+        pos_query = self._query_dense(rel_embeddings)
+        pos_key = self._key_dense(rel_embeddings)
+
+        # c2p
+        c2p_attn_scores = tf.einsum(
+            "aecd,abcd->acbe",
+            pos_key,
+            query,
+        )
+        c2p_pos = tf.clip_by_value(
+            rel_pos + rel_attn_span, 0, rel_attn_span * 2 - 1
+        )
+        c2p_pos = tf.broadcast_to(
+            c2p_pos,
+            shape=(
+                batch_size,
+                self.num_heads,
+                num_positions,
+                num_positions,
+            ),
+        )
+
+        c2p_attn_scores = tf.gather(
+            c2p_attn_scores,
+            indices=c2p_pos,
+            batch_dims=3,
+        )
+        c2p_attn_scores = tf.multiply(c2p_attn_scores, self.scale_factor)
+        score += c2p_attn_scores
+
+        # p2c
+        p2c_attn_scores = tf.einsum(
+            "aecd,abcd->acbe",
+            pos_query,
+            key,
+        )
+        p2c_pos = tf.clip_by_value(
+            -rel_pos + rel_attn_span, 0, rel_attn_span * 2 - 1
+        )
+        p2c_pos = tf.broadcast_to(
+            p2c_pos,
+            shape=(
+                batch_size,
+                self.num_heads,
+                num_positions,
+                num_positions,
+            ),
+        )
+        p2c_attn_scores = tf.gather(
+            p2c_attn_scores,
+            indices=p2c_pos,
+            batch_dims=3,
+        )
+        p2c_attn_scores = tf.transpose(p2c_attn_scores, [0, 1, 3, 2])
+        p2c_attn_scores = tf.multiply(p2c_attn_scores, self.scale_factor)
+        score += p2c_attn_scores
+
+        return score
+
+    def call(
+        self,
+        hidden_states,
+        rel_embeddings,
+        attention_mask=None,
+        return_attention_scores=False,
```

Comment:
I don't think we need this. Not being used that I can see.

Reply:
It is being used here:

```
        if return_attention_scores:
            return attention_output, attention_scores
```

The reason why I have kept this is so that this is consistent with the MHA layer.